tachyon/crypto/hashes/sponge/poseidon2/poseidon2.h

```diff
@@ -58,10 +59,20 @@ struct Poseidon2Sponge final
     if (is_full_round) {
       ExternalMatrix::Apply(state.elements);
     } else {
-      if constexpr (F::Config::kModulusBits <= 32) {
-        if (config.internal_diagonal_minus_one.rows() == 0) {
-          Poseidon2Plonky3InternalMatrix<F>::Apply(state.elements,
-                                                   config.internal_shifts);
+      using PrimeField =
+          std::conditional_t<math::FiniteFieldTraits<F>::kIsPackedPrimeField,
+                             typename math::FiniteFieldTraits<F>::PrimeField,
+                             F>;
+
+      if constexpr (PrimeField::Config::kModulusBits <= 32) {
+        if (config.use_plonky3_internal_matrix) {
+          if constexpr (math::FiniteFieldTraits<F>::kIsPackedPrimeField) {
+            Poseidon2Plonky3InternalMatrix<F>::Apply(
+                state.elements, config.internal_diagonal_minus_one);
+          } else {
+            Poseidon2Plonky3InternalMatrix<F>::Apply(state.elements,
```

Comment:
I think you should use PrimeField instead of F here.

Reply:
I think it's fine because that must be `PrimeField` due to the above condition.